### PR TITLE
Add try_cast function

### PR DESCRIPTION
**Pull request by kovrus**

### CODE REVIEW

File: sql/src/main/java/io/crate/operation/scalar/cast/CastFunctionResolver.java [hunk 1]
```diff
@@ -88,14 +93,30 @@
...
+    public static boolean supportsExplicitConversion(DataType returnType) {
+        return functionMap.get(returnType) != null;
+    }
+
+    public static Map<DataType, String> tryFunctionsMap() {
+        if (tryFunctionsMap != null) {
+            return tryFunctionsMap;
+        }
+        tryFunctionsMap = new HashMap<>();
+        for (Map.Entry<DataType, String> function : functionMap.entrySet()) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/scalar/cast/CastFunctionResolver.java]: return a view instead of creating a copy ... you can use Maps.transformValues for that

File: sql/src/main/java/io/crate/operation/scalar/cast/TryScalarFunction.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        }
+        return symbol;
+    }
+
+    @Override
+    public Object evaluate(Input[] args) {
+        assert args.length == 1;
+        try {
+            return returnType.value(args[0].value());
+        } catch (Exception e) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/scalar/cast/TryScalarFunction.java]: can we now be more explicit here ? only catch ClassCastException | IllegalArgumentException | ConversionException

File: sql/src/main/java/io/crate/operation/scalar/cast/TryScalarFunction.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+import io.crate.operation.scalar.ScalarFunctionModule;
+import io.crate.planner.symbol.Function;
+import io.crate.planner.symbol.Literal;
+import io.crate.planner.symbol.Symbol;
+import io.crate.types.DataType;
+
+import java.util.List;
+import java.util.Map;
+
+public class TryScalarFunction extends Scalar<Object, Object> {
```
Comment (mfelsche) [inline sql/src/main/java/io/crate/operation/scalar/cast/TryScalarFunction.java]: naming: TryCastScalarFunction

File: sql/src/test/java/io/crate/analyze/SelectStatementAnalyzerTest.java [hunk 4]
```diff
@@ -1468,6 +1467,42 @@ public void testCastExpression() throws Exception {
...
+                Arrays.<DataType>asList(DataTypes.LONG)));
+
+        analysis = analyze("select try_cast(1+1 as string) from users");
+        assertThat(analysis.relation().querySpec().outputs().get(0), isLiteral("2", DataTypes.STRING));
+
+        analysis = analyze("select try_cast(counters as array(boolean)) from users");
+        assertThat(analysis.relation().querySpec().outputs().get(0), isFunction(
+                CastFunctionResolver.tryFunctionsMap().get(new ArrayType(DataTypes.BOOLEAN)),
+                Arrays.<DataType>asList(new ArrayType(DataTypes.LONG))));
+        analysis = analyze("select try_cast(1+1 as string) from users");
```
Comment (mfelsche) [inline sql/src/test/java/io/crate/analyze/SelectStatementAnalyzerTest.java]: do some asserts on the result

**Comment by mfelsche** [comment]
add some tests with null values

otherwise ack
